### ISSUE TITLE
easily specify container image versions

### ISSUE THREAD

**Issue by jbianquetti-nami** [bitnami/kube-prod-runtime#340]
Otherwise, you'll need to create a bunch of boilerplate like this:

```
    // BKPR: prometheus.nodeExporter
    nodeExporter+: {
      // BKPR: prometheus.nodeExporter.daemonset (DaemonSet)
      daemonset+: {
        spec+: {
          template+: {
            spec+: {
              containers_+: {
                default+: {
                  // Match the version we're running on other clusters
                  image: "bitnami/node-exporter:0.16.0",
                },
              },
            },
          },
        },
      },
},
```

**Comment by anguslees** [comment]
I feel like changing the image often goes with other changes (you want a new image so you can enable a new command line option, for example).

If we build an "easy" way to specify image versions (eg: expose `images.json` at the top-level of components), then we're really just setting an anti-pattern for other customisations.  We can't move "everything" up to the top-level (the k8s resource definition is in a deeply-nested form due to legitimate reasons).

I guess I want to see that modifying the images is so common that it demands special treatment before committing to support a deviation from the usual k8s api structure.  .. and I don't think we can state that until we have more miles under the current (standard k8s api) structure.

**Comment by anguslees** [comment]
Note: you could also improve this locally with your own helper function.  Something like:

```javascript
local overrideImage(image, container="default") = {
  spec+: {
    template+: {
      spec+: {
        containers_+: {
          [container]+: {
            image: image,
          },
        },
      },
    },
  },
}

{
  nodeExporter+: {
    // Match the version we're running on other clusters
    daemonset+: overrideImage("bitnami/node-exporter:0.16.0"),
  },
}
```

Again, I would advise _not_ using this as a generic pattern for all possible overrides (the scope is too broad, and you'll just be writing/calling functions everywhere).

**Comment by falfaro** [comment]
Another option would be to import/load the `images.json` file from the root manifest, where things can be overridden if necessary, and propagate this object to the underlying platform manifest (e.g. `platforms/eks.jsonnet`). At the moment, each manifest imports `images.json` which makes image overrides complicated. Something like:

```
// Cluster-specific configuration
(import "file:///Users/falfaro/go/src/github.com/bitnami/kube-prod-runtime/manifests/platforms/eks.jsonnet") {
  config:: import "kubeprod-autogen.json",
  images:: (import "../components/images.json") {
    alertmanager: "bitnami/alertmanager:0.16.2-r12",
  }
}
```

**Comment by anguslees** [comment]
> Another option would be to import/load the images.json file from the root manifest, where things can be overridden if necessary

Right, I this this is the specific change we're considering here.  Internally we would need to pass the `images` object down to each component within the platform jsonnet file, because each component is nested, but that's something we can manage.

My main concern is it doesn't actually make anything substantially "easier".  It's still _conceptually_ one "line" to override the image with any approach: The deeply nested version is more than one line of actual text, but it's still conceptually a single "set this parameter to this value" action.
And the downside with any deviation from the standard k8s structure is that it is surprising and new.  We also then need BKPR-specific docs and howtos to explain it rather than just "read the standard k8s resource docs".

TL;DR: the fix I would _like_ to see here is a jsonnet syntax improvement https://github.com/google/jsonnet/issues/227

**Comment by jbianquetti-nami** [comment]
> TL;DR: the fix I would like to see here is a jsonnet syntax improvement google/jsonnet#227

That's would be amazingly simple, but seems like there are no movements on the PR for a few months.

**Comment by arapulido** [comment]
@jjo is this fixed with your PR with the helper.jsonnet?

**Comment by jjo** [comment]
It'd be simplified by it (with a different approach), I think we may
also want to add `overrideImage()` from Gus above to `helpers.jsonnet`.

**Comment by anguslees** [comment]
See also https://github.com/anguslees/kustomize-libsonnet for a kustomize-inspired approach.
The scope for potential "helpers"  is unlimited :)
